File: src/cli/commands/session.ts
import { BrowserPlatformError } from '../../core/errors.js';
import { requireFlag } from '../argv.js';
import { actInSession, closeSession, getSessionContext, observeSession, openSession, snapshotSession } from '../../daemon/client.js';
import { handleDaemonEnsure } from './daemon.js';
import type { SessionActionPayload, SessionBackend } from '../../daemon/types.js';

export async function handleSessionOpen(args: string[]): Promise<unknown> {
  await handleDaemonEnsure();
  const url = requireFlag(args, '--url');
  const storageStateIndex = args.indexOf('--storage-state');
  const storageStatePath =
    storageStateIndex !== -1 && storageStateIndex < args.length - 1 ? args[storageStateIndex + 1] : undefined;
  const backend = resolveBackend(args);
  return openSession(url, { storageStatePath, backend });
}

export function resolveBackend(args: string[]): SessionBackend {
  const backendIndex = args.indexOf('--backend');
  if (backendIndex === -1) {
    return 'chromium';
  }

  if (backendIndex >= args.length - 1) {
    throw new BrowserPlatformError('--backend requires a value. Allowed values: chromium, camoufox', {
      code: 'INVALID_BACKEND'
    });
  }

  const backendRaw = args[backendIndex + 1]?.toLowerCase();
  if (backendRaw === 'chromium' || backendRaw === 'camoufox') {
    return backendRaw;
  }

  throw new BrowserPlatformError('Unsupported backend. Allowed values: chromium, camoufox', {
    code: 'INVALID_BACKEND'
  });
}

export async function handleSessionContext(args: string[]): Promise<unknown> {
  const sessionId = requireFlag(args, '--session');
  return getSessionContext(sessionId);
}

export async function handleSessionObserve(args: string[]): Promise<unknown> {
  const sessionId = requireFlag(args, '--session');
  return observeSession(sessionId);
}

export async function handleSessionAct(args: string[]): Promise<unknown> {
  const sessionId = requireFlag(args, '--session');
  const json = requireFlag(args, '--json');

  let payload: SessionActionPayload;
  try {
    payload = JSON.parse(json) as SessionActionPayload;
  } catch (error) {
    throw new BrowserPlatformError('Invalid action payload JSON', {
      code: 'INVALID_JSON_PAYLOAD',
      details: { cause: error instanceof Error ? error.message : String(error) }
    });
  }

  if (!payload || typeof payload !== 'object' || !('action' in payload)) {
    throw new BrowserPlatformError('Action payload must contain an action field', { code: 'INVALID_ACTION_PAYLOAD' });
  }

  return actInSession(sessionId, payload);
}

export async function handleSessionSnapshot(args: string[]): Promise<unknown> {
  const sessionId = requireFlag(args, '--session');
  return snapshotSession(sessionId);
}

export async function handleSessionClose(args: string[]): Promise<unknown> {
  const sessionId = requireFlag(args, '--session');
  return closeSession(sessionId);
}

File: browser_platform/cli/commands/session.py
from __future__ import annotations
import json

from ...core.errors import BrowserPlatformError
from ...daemon.types import SESSION_BACKENDS
from ...daemon.client import (
    act_in_session,
    close_session,
    get_session_context,
    observe_session,
    open_session,
    snapshot_session,
)
from ..argv import require_flag, optional_flag
from .daemon import handle_daemon_ensure

_ALLOWED_BACKENDS_TEXT = ', '.join(SESSION_BACKENDS)


async def handle_session_open(args: list[str]) -> dict:
    await handle_daemon_ensure()
    url = require_flag(args, '--url')
    storage_state_path = optional_flag(args, '--storage-state')
    profile_id = optional_flag(args, '--profile')
    scenario_id = optional_flag(args, '--scenario')
    backend = _resolve_backend(args)
    return await open_session(
        url,
        storage_state_path=storage_state_path,
        backend=backend,
        profile_id=profile_id,
        scenario_id=scenario_id,
    )


def _resolve_backend(args: list[str]) -> str:
    try:
        index = args.index('--backend')
    except ValueError:
        return 'camoufox'

    if index >= len(args) - 1:
        raise BrowserPlatformError(
            f'--backend requires a value. Allowed values: {_ALLOWED_BACKENDS_TEXT}',
            code='INVALID_BACKEND',
        )
    raw = args[index + 1].lower()
    if raw in SESSION_BACKENDS:
        return raw
    raise BrowserPlatformError(
        f'Unsupported backend. Allowed values: {_ALLOWED_BACKENDS_TEXT}',
        code='INVALID_BACKEND',
    )


async def handle_session_context(args: list[str]) -> dict:
    session_id = require_flag(args, '--session')
    return await get_session_context(session_id)


async def handle_session_observe(args: list[str]) -> dict:
    session_id = require_flag(args, '--session')
    return await observe_session(session_id)


async def handle_session_act(args: list[str]) -> dict:
    session_id = require_flag(args, '--session')
    raw_json = require_flag(args, '--json')
    try:
        payload = json.loads(raw_json)
    except Exception as exc:
        raise BrowserPlatformError(
            'Invalid action payload JSON',
            code='INVALID_JSON_PAYLOAD',
            details={'cause': str(exc)},
        ) from exc

    if not isinstance(payload, dict) or 'action' not in payload:
        raise BrowserPlatformError(
            'Action payload must contain an action field', code='INVALID_ACTION_PAYLOAD'
        )
    return await act_in_session(session_id, payload)


async def handle_session_snapshot(args: list[str]) -> dict:
    session_id = require_flag(args, '--session')
    return await snapshot_session(session_id)


async def handle_session_close(args: list[str]) -> dict:
    session_id = require_flag(args, '--session')
    return await close_session(session_id)
